scrapy/resolver.py

```diff
@@ -1,19 +1,38 @@
 from twisted.internet import defer
 from twisted.internet.base import ThreadedResolver
+from twisted.internet.interfaces import IHostnameResolver, IResolutionReceiver, IResolverSimple
+from zope.interface.declarations import implementer, provider
 
 from scrapy.utils.datatypes import LocalCache
 
-# TODO: cache misses
 
+# TODO: cache misses
 dnscache = LocalCache(10000)
 
 
+@implementer(IResolverSimple)
 class CachingThreadedResolver(ThreadedResolver):
+    """
+    Default caching resolver. IPv4 only, supports setting a timeout value for DNS requests.
+    """
+
     def __init__(self, reactor, cache_size, timeout):
         super(CachingThreadedResolver, self).__init__(reactor)
         dnscache.limit = cache_size
+        self.reactor = reactor
```

Comment:
I think self.reactor is already set in the base class here

Reply:
Right, I forgot about checking what the parent class was doing. Indeed, [the reactor is stored as an attribute](https://github.com/twisted/twisted/blob/twisted-19.10.0/src/twisted/internet/base.py#L243). Thanks Mikhail!